File: permutations.erl
%%%--------------------------------------------------------------------------------------------------%%%
%%% @author Shahzad Bhatti
%%% @doc Generate permutations of string
%%%     erlc perms.erl
%%%     erl -noshell -s permutations test_perms -s init stop
%%%--------------------------------------------------------------------------------------------------%%%

-module(permutations).
-export([perms/1, test_perms/0]).

perms([]) ->
    [" "];
perms([H|T]) -> 
    Words = perms(T),
    make_words(H, Words, []).

make_words(First, Words, Output) ->
    lists:foldl(fun(Word, OutputList) -> add_words(First, Word, 1, OutputList) end, Output, Words).

add_words(_C, Word, I, Output) when I == length(Word) + 1 ->
    Output;
add_words(C, [], _I, Output) ->
    Output;
add_words(C, Word, I, Output) ->
    Pre = lists:sublist(Word, 1, I),
    Suf = lists:sublist(Word, I+1, length(Word)-I),
    Word1 = Pre ++ [C|Suf],
    add_words(C, Word, I+1, [Word1|Output]).

test_perms() ->
    List = perms("abc"),
    lists:foreach(fun(Word) -> io:format("~w\n", [list_to_atom(Word)]) end,
        List).

File: permutations.rb
###--------------------------------------------------------------------------------------------------###
### @author Shahzad Bhatti
### @doc Generate permutations of a string
###--------------------------------------------------------------------------------------------------###
 
# Permutations = N!
# N Pick K = nPk (5 people sit on 3 chairs) = N! / (N-K)!  
# e.g. 3 letters in alphabets (26) = 26 ** 3
# e.g. 3 letters in alphabets if we need different letters = 26! / (26-3)!
# color code game - blue/yello/white/red/orange/green, pick 4 colors if colors
# cannot be repeated: 6!/(6-4)!
# combinations: How many ways to choose combinations (divide by # of ways you
# can arrange k things)
# n Choose k  = nCk = n!/((n-k)! * k!) -- binomial coefficient -- order doesn't matter
# e.g. handshakes with 4 people = 4C2 = ( 4 2 ) = 4! / ((4-2)! * 2!) = 6
# card game using 36 cards, 4 suits, hand is collection of 9, how many
# hands possible = ( 36 9) = 36! / ((36 - 9)! * 9!) = 
#
def perms(str)
    return [""] if str == nil || str.length == 0
    first = str.slice(0, 1)
    words = perms(str.slice(1,str.length))
    perms = []
    for word in words
        (0..word.length).each do |j|
            perms << word.slice(0,j) + first + word.slice(j,word.length)
        end
    end
    perms
end

puts perms("abc").inspect
